src/devices/cddadevice.cpp

```diff
@@ -73,12 +71,23 @@ void CddaDevice::WatchForDiscChanges(bool watch) {
     disc_changed_timer_.stop();
 }
 
-void CddaDevice::LoadSongs() { cdda_song_loader_.LoadSongs(); }
+void CddaDevice::LoadSongs() {
+  cdda_song_loader_.LoadSongs();
+  disc_changed_timer_.stop();
+}
 
 void CddaDevice::SongsLoaded(const SongList& songs) {
   model_->Reset();
-  emit SongsDiscovered(songs);
   song_count_ = songs.size();
+  emit SongsDiscovered(songs);
+}
+
+void CddaDevice::SongsLoadingFinished() {
+  // Ensure that cdio_get_media_changed is cleared
+  // after songs are loaded, so we don't potentially
+  // re-read the same disc.
+  cdio_get_media_changed(cdio_);
```

Comment:
The problem I was trying to address with this is that cdio_get_media_changed is returns true twice when a disc is inserted, with a bit of delay between both (I think they roughly correspond to "tray is closed" and "disc was detected/read/spun up/whatever"). There's no way to programmatically distinguish them and for the current code this means that the first time cdio_get_media_changed "fires" we start reading the disc, deactivating the disc change polling timer while doing so. After reading is complete, the timer gets reactivated and, due to the behavior described above, sees cdio_get_media_changed returned true, causing another reading of the disc. Therefore I decided to "clear" the cdio_get_media_changed response after reading to avoid this.

I can see your concern with what happens if the media changes during the load process. I tried to provoke this, but I was unable to as the reading is quite fast and I think the tray is locked while that is in progress. I could only get the drive to eject the disc before the read started (while it was still spinning up) or after the read completed (with a bit of a delay for spin down). I think it is near impossible to actually have a disc change in the time between the read completing and this reset of cdio_get_media_changed occuring and it wouldn't be a huge problem anyways (user would just have to cycle the tray once in case it happens).

That said, it just occured to me that I still need to check this without an internet connection since waiting for a response from musicbrainz may delay this long enough for a physical disc change to actually take place...